### PR TITLE
arm GIC V3 driver

### PR DESCRIPTION
**Pull request by sandeepbrcm**
Adds basic driver for GIC-V3 with system interface enabled.
- Patch is tested on Broadcom SOC with Cortex A72 and GIC 500. #24633 
1- GIC V2 Legacy mode is not supported by the hardware by configuration hence it was not attempted to be tested with this driver. Also it is deprecated feature.

2- This driver currently supports Secure EL1 only. Also assumes it is the owner of all interrupts. However NS EL1 can be easily added in future.
- If Zephyr is run along with a EL3 runtime with interrupt support then there needs to be more work to  define ownership of Group 0 and Group 1 Secure interrupts.

3- There are many TODOs in GICV3 but I think this is a logically complete functional drop.

4- TODO: GICV2 GICV3 common code ?  This can be explored in stages. Some of it I can make in this PR if required.  In my view keeping them separate but identical may be better in terms of maintenance.  One reason I did not attempt first is I do not have a GICV2 hardware  and do not want to risk with Qemu behavior.

### CODE REVIEW

File: drivers/interrupt_controller/intc_gicv3.c [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+#define ICC_RPR_EL1             S3_0_C12_C11_3
+#define ICC_IGRPEN1_EL3         S3_6_c12_c12_7
+#define ICC_IGRPEN0_EL1         S3_0_c12_c12_6
+#define ICC_HPPIR0_EL1          S3_0_c12_c8_2
+#define ICC_HPPIR1_EL1          S3_0_c12_c12_2
+#define ICC_IAR0_EL1            S3_0_c12_c8_0
+#define ICC_IAR1_EL1            S3_0_c12_c12_0
+#define ICC_EOIR0_EL1           S3_0_c12_c8_1
+#define ICC_EOIR1_EL1           S3_0_c12_c12_1
+#define ICC_SGI0R_EL1           S3_0_c12_c11_7
```
Comment (sandeepbrcm) [inline drivers/interrupt_controller/intc_gicv3.c:70]: Since these are cpu core system registers I am moving these register defines and their access macro to cpu.h  and moving to another PR.

File: drivers/interrupt_controller/intc_gicv3.c [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+{
+	u32_t mask = BIT(intid & NUM_INTR_PER_REG);
+	u32_t idx = intid / 32;
+	mem_addr_t base;
+
+	base = (intid < GIC_SPI_BASE) ?
+	       (GIC_RDIST_BASE + GICR_SGIBASE_OFFSET)
+	       : GIC_DIST_BASE;
+
+	sys_write32(~mask, ICENABLER(base, idx));
```
Comment (Abhishek-brcm) [inline drivers/interrupt_controller/intc_gicv3.c:160]: Is this supposed to be `~mask`? Since this is clear enable register, do we have to directly write `mask` here?

File: drivers/interrupt_controller/intc_gicv3.c [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+#define ICC_IGRPEN0_EL1         S3_0_c12_c12_6
+#define ICC_HPPIR0_EL1          S3_0_c12_c8_2
+#define ICC_HPPIR1_EL1          S3_0_c12_c12_2
+#define ICC_IAR0_EL1            S3_0_c12_c8_0
+#define ICC_IAR1_EL1            S3_0_c12_c12_0
+#define ICC_EOIR0_EL1           S3_0_c12_c8_1
+#define ICC_EOIR1_EL1           S3_0_c12_c12_1
+#define ICC_SGI0R_EL1           S3_0_c12_c11_7
+
+#define	GIC_SPI_BASE		32
```
[the remark this answers is not in the document]
Comment (sandeepbrcm) [reply drivers/interrupt_controller/intc_gicv3.c:72]: I felt the `INT`  in the name was not necessary as it is `Shared Peripheral Interrupt` .  But since it is already in use I can reuse it.

File: drivers/interrupt_controller/intc_gicv3.c (hunk 2 above)
Comment (sandeepbrcm) [reply drivers/interrupt_controller/intc_gicv3.c:160]: Thanks.  It must be `mask`   for` ICENABLER `.  Since we never do a `irq_disable(irq) ` we didn't see any impact.

File: drivers/interrupt_controller/intc_gicv3.c [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+/* Initialize the cpu interface */
+static void gicv3_cpuif_init(void)
+{
+	u32_t icc_sre;
+	u32_t intid;
+
+	mem_addr_t base = (GIC_RDIST_BASE + GICR_SGIBASE_OFFSET);
+
+	/* Disable all sgi ppi */
+	sys_write32(~0, ICENABLER((GIC_RDIST_BASE + GICR_SGIBASE_OFFSET), 0));
```
[the remark this answers is not in the document]
Comment (sandeepbrcm) [reply drivers/interrupt_controller/intc_gicv3.c:222]: ACK

File: drivers/interrupt_controller/intc_gicv3.c [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+
+	num_ints = sys_read32(base + GIC_TYPER);
+	num_ints &= 0x1f;
+	num_ints = (num_ints + 1) << 5;
+
+	/*
+	 * Default configuration of all SPIs
+	 */
+	for (intid = GIC_SPI_BASE; intid < num_ints; intid += 32) {
+		idx = intid / NUM_INTR_PER_REG;
```
[the remark this answers is not in the document]
Comment (sandeepbrcm) [reply drivers/interrupt_controller/intc_gicv3.c:286]: Good catch.  Gone bad in my cleanup. And tests pass because by chance all` idx` are getting selected. 
I replaced `32` with `NUM_INTR_PER_REG`.  The define name I used is wrong.
Other places I am using ` (intid & NUM_INTR_PER_REG)  `to extract the reminder.  Will correct in next version.

File: drivers/interrupt_controller/intc_gicv3.c [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+		/* Clear pending */
+		sys_write32(0xffffffff, ICPENDR(base, idx));
+		/* All SPIs are G1S and owned by Zephyr */
+		sys_write32(0, IGROUPR(base, idx));
+		sys_write32(0xffffffff, IGROUPMODR(base, idx));
+	}
+
+	/* Configure default priorities for SPI */
+	for (intid = GIC_SPI_BASE; intid < num_ints; intid += 4) {
+		sys_write32(0xa0a0a0a0, IPRIORITYR(base, intid));
```
[the remark this answers is not in the document]
Comment (sandeepbrcm) [reply drivers/interrupt_controller/intc_gicv3.c:298]: Kept it same as` NS` state value.  But really doesn't matter as I have kept `PMR` as` 0xff` allowing all priorities.
I can add appropriate define.

**Comment by sandeepbrcm** [comment]
Sorry . working dir is messed up. Will abandon this PR and raise new one.
